packages/doenetml-worker/lib-doenetml-core/src/dast/ref_resolve.rs

```diff
@@ -128,13 +148,20 @@ pub struct Resolver {
     /// `name_map[0]` is a map of all names accessible from the `flat_root`
     /// and `name_map[i+1]` is a map of all names accessible from node `i`.
     name_map: Vec<HashMap<String, Ref>>,
+    /// Map of resolutions of indices that follow a match to a node, where only indices of `index_resolutions` itself are shifted by `1`.
+    /// Given a path `initial_part[i+1]` where `initial_part` has matched node `j`,
+    /// we look up `index_resolutions[j+1][i]` to determine if the (already shifted) index `i+1` matches a node.
+    /// If `index_resolutions[j+1][i] = Some(k)`, then the path `initial_part[i+1]` resolves to node `k`.
+    /// Note: `index_resolutions[j+1][i] = None` indicates that `initial_part[i+1]` would resolve to a text node,
+    /// which is not a valid ref resolution.
+    index_resolutions: Vec<Vec<Option<Index>>>,
```

Comment:
We now have a bunch of vectors of the same length hanging around. I think we got here by adding them one by one, but I'm not sure if its serving us well anymore.

So, every node as a `parent`, `resolution_algorithm`, `name_map`, and now `index_resolutions`. I think we can make a datastructure with all of those and have the Resolver just work on a `Vec` of that structure.

There should also be less talking about `+1` on indices, I think...